### PR TITLE
os/arch/arm/src, os/board/rtl8730e: SMP for RTL8730E

### PR DESCRIPTION
**Pull request by edwakuwaku**
- SMP dual core feature porting
- Verified running ok with smp_main.c (Test on XIP)
- Wifi basic function ok
- BT init working fine
- Hotplug mode implementation will have an independent PR
- Ported customization for CPU gating when doing flash operation

### CODE REVIEW

File: os/arch/arm/src/armv7-a/arm_cpuhead.S
Comment (edwakuwaku) [inline os/arch/arm/src/armv7-a/arm_cpuhead.S:248]: This cache property is not needed to be enabled. Will remove it.

**Review by r-prabu — APPROVED** [review]

File: os/arch/arm/src/armv7-a/arm_cpuhead.S
Comment (abhishek-samsung) [inline os/arch/arm/src/armv7-a/arm_cpuhead.S:252]: Why is the change necessary? We will be only using ttbr0 so ttbr1 can be set regardless.

**Review by abhishek-samsung — APPROVED** [review]

File: os/arch/arm/src/armv7-a/arm_cpuhead.S
Comment (edwakuwaku) [reply os/arch/arm/src/armv7-a/arm_cpuhead.S:252]: I think that since we are not utilizing ttbr1, it is meaningless to set it. If you think this change is redundant, I can remove it.

Comment (abhishek-samsung) [reply os/arch/arm/src/armv7-a/arm_cpuhead.S:252]: Yes, please revert it if it doesn't cause any issue. Lets only necessary changes under config checks.

Comment (edwakuwaku) [reply os/arch/arm/src/armv7-a/arm_cpuhead.S:252]: Hello @abhishek-samsung ,

Okay, I will make changes to all the comment once power save PR is merged. As there is dependency in between these two, a rebase is needed for SMP. In the meantime, you may review and leave comment based on your concern.

File: os/arch/arm/Kconfig
Comment (sunghan-chang) [inline os/arch/arm/Kconfig:276]: Could you let me know why you add this?

File: os/arch/arm/src/armv7-a/arm_cpuflash.c
Comment (sunghan-chang) [inline os/arch/arm/src/armv7-a/arm_cpuflash.c:19]: Let's modify or remove the path

File: os/arch/arm/src/armv7-a/arm_cpuflash.c
Comment (sunghan-chang) [inline os/arch/arm/src/armv7-a/arm_cpuflash.c:21]: Do you get this from Nuttx? If not, it looks no need to add this license.

**Comment by sunghan-chang** [comment]
```
build/configs/rtl8730e: Place secondary bootup code in PSRAM
```
Could you leave why in the commit description? Same as other commits.
And minor typo, booup -> bootup

File: build/configs/rtl8730e/loadable_apps_16M/defconfig
Comment (edwakuwaku) [reply build/configs/rtl8730e/loadable_apps_16M/defconfig:1407]: This is to sync the changes according to Kconfig, is it not needed?

**Comment by edwakuwaku** [comment]
> ```
> build/configs/rtl8730e: Place secondary bootup code in PSRAM
> ```
> 
> Could you leave why in the commit description? Same as other commits. And minor typo, booup -> bootup

Yes, I will add in the commit description. The main reason is because when we boot the secondary core, the entry address to jump from ATF BL32 to BL33 is set to PSRAM address, so we move only the startup codes which is the early stage boot up code in arm_cpuhead.c to PSRAM.
I didn't see any typo in the commit message.

File: os/arch/arm/src/armv7-a/arm_cpuflash.c
Comment (edwakuwaku) [reply os/arch/arm/src/armv7-a/arm_cpuflash.c:21]: No, this is a customisation for AIDual, we registered an additional software generated interrupt (SGI) to handle flash operation.

File: os/arch/arm/Kconfig
Comment (edwakuwaku) [reply os/arch/arm/Kconfig:276]: There are 2 reasons for adding this macro:
1) This is one of the dependency macro to enable CONFIG_SMP, thus it has to be enabled to have SMP appearing in the menuconfig selection
2) This will enable the spinlock mechanism to handle multicore coherency, you can view it from the file : os/arch/arm/src/armv7-a/arm_testset.S

**Comment by sunghan-chang** [comment]
The typo is in the 2nd commit title.

File: os/arch/arm/src/armv7-a/arm_cpuflash.c
Comment (edwakuwaku) [reply os/arch/arm/src/armv7-a/arm_cpuflash.c]: done~

File: os/arch/arm/src/armv7-a/arm_cpuflash.c
Comment (edwakuwaku) [reply os/arch/arm/src/armv7-a/arm_cpuflash.c]: done~

File: os/arch/arm/src/armv7-a/arm_cpuflash.c
Comment (edwakuwaku) [reply os/arch/arm/src/armv7-a/arm_cpuflash.c]: done~

**Review by sunghan-chang — APPROVED** [review]

**Comment by sunghan-chang** [comment]
@kishore-sn @r-prabu Could you review and leave an approval?

**Comment by edwakuwaku** [comment]
@sunghan-chang  @kishore-sn , are we merging this?

**Comment by sunghan-chang** [comment]
@edwakuwaku I heard there is a crash with SMP enabled. After merge this, if we don't enable loadable or smp, is it ok to run non-smp (current usage)?

**Comment by sunghan-chang** [comment]
@edwakuwaku Even it has an issue internally, before we fix it, do you want to merge this now?

**Comment by edwakuwaku** [comment]
@sunghan-chang , the crash Is happening in loadable apps config with SMP enabled, and there are 2 purpose for this PR:
1) Update RTK SMP functionality
2) Fix NUTTX cortex-A code logics
Besides, the SMP is disabled currently in all the defconfigs, so merging this PR should not have conflict with the crash problem, since we are still running in single core. And even if the solution is found for loadable crash, it should be raised in another PR.

**Review by sunghan-chang — APPROVED** [review]

**Pull request merged**
